### DISCUSSION TITLE
✨ Support pydantic model as individual cookie type

### DISCUSSION THREAD

**Discussion by jd-solanki** [fastapi/fastapi#12196] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```py
from typing import Annotated

from fastapi import Cookie, FastAPI
from pydantic import BaseModel

app = FastAPI()


class DataConfigCookie(BaseModel):
    name: str


class Cookies(BaseModel):
    data_config: DataConfigCookie


"""
* Set cookie in browser console:

function setCookie(name, value, days) {
  const expires = new Date(Date.now() + days * 24 * 60 * 60 * 1000).toUTCString();
  document.cookie = `${name}=${encodeURIComponent(JSON.stringify(value))}; expires=${expires}; path=/`;
}

setCookie("data_config", {name: "data_config"}, 1);
"""


@app.get("/")
def read_root(
    cookies: Annotated[Cookies, Cookie()],
):
    data_config = cookies.data_config
    print(f"data_config_cookie: {data_config}")
    return {"Hello": "World"}

```

<details>
<summary>Old</summary>

```python
from typing import Annotated

from fastapi import Cookie, FastAPI
from pydantic import BaseModel

app = FastAPI()


class MyCookie(BaseModel):
    name: str


@app.get("/")
def read_root(
    data_config: Annotated[MyCookie, Cookie()],
):
    print(f"data_config_cookie: {data_config}")
    return {"Hello": "World"}
```

</details>

### Description

When we run it we get validation error on endpoint:
```
Input should be a valid dictionary or object to extract fields from
```

---

It'll suite the DX of FastAPI if it can autoparse the JSON cookie & throw pydantic error accordingly.

Right now we have to define type as `str` and manually parse it and we encounter internal server error if validation fails so we also have to handle pyndatic validation.

### Operating System

Linux

### Operating System Details

Ubuntu 20.04

### FastAPI Version

0.115.0

### Pydantic Version

2.8.2

### Python Version

Python 3.12.5

### Additional Context

```py
# data_config: Annotated[str, Cookie()]

decoded_cookie = unquote(data_config)
config = MyCookie.model_validate_json(decoded_cookie)
```

Parsing like this raises internal server error if validation fails so we have to handle it manually as well.

**Comment by jd-solanki** [comment]
~~Resolved by new release: [0.115.0](https://github.com/fastapi/fastapi/releases/tag/0.115.0)~~

**Reply by jd-solanki** [reply]
Unfortunetly, It doesn't solve cookie as object

**Comment by yakubka** [comment]
FastAPI does not natively support deserializing a cookie value into a Pydantic model because cookies are plain strings in HTTP and there is no standard encoding for structured data. What you can do is parse the cookie value manually inside a dependency.

```python
from fastapi import Cookie, Depends
from pydantic import BaseModel
import json

class SessionData(BaseModel):
    user_id: int
    role: str

def get_session(session: str | None = Cookie(default=None)) -> SessionData | None:
    if session is None:
        return None
    try:
        return SessionData.model_validate_json(session)
    except Exception:
        return None

@app.get("/me")
async def me(session: SessionData | None = Depends(get_session)):
    if session is None:
        raise HTTPException(status_code=401)
    return session
```

The cookie value must be JSON-encoded when set, and the dependency decodes it. This gives you Pydantic validation on read while keeping the cookie as a plain string on the wire.
